Add a transformer neural operator model and an accompanying example for training it on Darcy flow

zijieli-Jlee opened the pull request:
Main changes:

* Add `transformer_no.py` under `models`, which is a Transformer encoder-decoder architecture for learning operator.  Small changes to the `__init__.py` and `model_dispatcher.py` accordingly.
* Add Gaussian Fourier feature and Siren to the `embeddings.py` which is useful for building the neural field (query point embedding) in the Transformer decoder.
* Add a training example `train_darcy_transformer.py` under `scripts`, the corresponding config yaml file is `darcy_transformer_config.yaml`.
* Fixed a bug in previous attention layer test [#290](https://github.com/neuraloperator/neuraloperator/issues/290)

dhpitt reviewed (commented):
Thanks @zijieli-Jlee! This looks awesome. A few minor things:

1.  Thanks for fixing #290 ! Would it be OK to split up this PR into one that fixes the issue (which we can merge immediately) and one that implements the TransformerNO (which may take longer to go over)?
2. It would be great to add unit tests for the `TransformerNO` model, along the lines of `neuralop/models/tests/test_fno.py` (just verifying that the forward pass produces outputs that we would expect)
3. It might help keep the code flexible and easier to maintain if the `EncoderBlocks` was its own module in `neuralop.layers`, along the lines of the `FNOBlocks`.

dhpitt commented on scripts/train_darcy_transformer.py:
```diff
@@ -0,0 +1,210 @@
+model = get_model(config)
+
+
+class ModelWrapper(torch.nn.Module):
```
What's the function of the ModelWrapper class? It seems like it adds functionality to the Transformer NO's forward pass. Some documentation would be helpful.

mliuschi replied:
Is it possible to incorporate the things in `ModelWrapper` into the dataset instead? I think this has been the convention we have been taking in the library. @JeanKossaifi Feel free to correct me if I'm wrong.

mliuschi replied:
Is it possible to incorporate the things in `ModelWrapper` into the dataset instead? I think this has been the convention we have been taking in the library. @JeanKossaifi Feel free to correct me if I'm wrong.

dhpitt commented on neuralop/models/transformer_no.py:
```diff
@@ -0,0 +1,170 @@
+            self.enc_pos_emb_module = None
+            self.dec_pos_emb_module = None
+
+        self.encoder = nn.ModuleList(
```
nit: `FNOBlocks` takes `n_layers` as a parameter instead of exposing a moduleList within the `FNO` itself. It might be nice to have a similar convention here

zijieli-Jlee replied:
Got it;)

dhpitt reviewed (commented):
Thank you for the updates @zijieli-Jlee, the PR is in really good shape overall. I left a few small comments

zijieli-Jlee commented:
Hi @dhpitt , thanks for the comments!

1. I saw this was fixed at #295 , let me know if there is still any remaining issue
2. Definitely. I will add the test for the `TransformerNO` if its current structure looks fine (see the next point)
3. Thanks for the suggestion! I added `transformer_block` following `fno_block` to `layers`, the `transformer_block` contains implementation for `TransformerEncoderBlock` and `TransformerDecoderBlock`. I also added test for these modules to `test_transformer_block`

dhpitt commented:
@zijieli-Jlee thanks for the edits, your PR is in good shape. if you fix the conflicts we can re-run the tests and see if it's ready to go.

dhpitt commented:
Thanks for the fixes @zijieli-Jlee ! I fixed a one-line conflict with a model import since the last update to main. 

This PR is looking great overall, but I think your layers and util functions could use clearer documentation in the docstrings - see `neuralop/layers/integral_transform.py` for an example of informative docs.

zijieli-Jlee commented:
I added docstring to the transformer_block and some newly added positional encoding class. In addition I just spotted and fixed a bug in the normalization inside attention layer

dhpitt commented on neuralop/layers/transformer_block.py:
```diff
@@ -0,0 +1,277 @@
+                pos_src: torch.Tensor, grid point coordinates of shape [batch_size, num_src_grid_points, channels]
+                pos_emb_module: nn.Module, positional embedding module, by default None
+                pos_qry: torch.Tensor, grid point coordinates of shape [batch_size, num_sry_grid_points, channels],
+                         by default None and is set to pos_src
```
It would be useful to include the specific purpose of `pos_qry` as compared to pos_src - inputs to the query basis function

dhpitt commented on neuralop/layers/embeddings.py:
```diff
@@ -73,4 +73,173 @@ def apply_2d_rotary_pos_emb(t, freqs_x, freqs_y):
+
+# Gaussian random Fourier features
+# code modified from: https://github.com/ndahlquist/pytorch-fourier-feature-networks
+class GaussianFourierFeatureTransform(nn.Module):
```
nit: these comments probably belong in the docstring

dhpitt commented on neuralop/layers/embeddings.py:
```diff
@@ -73,4 +73,173 @@ def apply_2d_rotary_pos_emb(t, freqs_x, freqs_y):
+
+
+# SirenNet
+# code modified from: https://github.com/lucidrains/siren-pytorch/blob/master/siren_pytorch/siren_pytorch.py
```
same as above - cleaner to leave these comments in the docstring

dhpitt reviewed (commented):
This PR is looking great. My only comment is a small nitpick about putting the model source within docstrings. Otherwise I think this is ready to go.

dhpitt commented:
Looks great! Asking @JeanKossaifi  for final approval.

mliuschi commented on neuralop/layers/transformer_block.py:
```diff
@@ -0,0 +1,282 @@
+                                     pos_src=pos,
+                                     positional_embedding_module=pos_emb_module,
+                                     **kwargs)
+            u = u + u_attention_skip
```
Have you noticed any empirical improvements from having this skip connection be a pointwise linear (not MLP) as opposed to identity connection? If so, it could be interesting to supplement with this functionality.

mliuschi reviewed (commented):
Left a couple of comments only because it mostly looks good to me!

juliusberner commented on neuralop/layers/embeddings.py:
```diff
@@ -73,4 +74,170 @@ def apply_2d_rotary_pos_emb(t, freqs_x, freqs_y):
+            Parameters:
+                dim_in: int, Number of input channels.
+                dim_out: int, Number of output channels.
+                w0: float, scaling factor (denominator) used to initialize the weights, by default 6.
```
`w0` seems to default to `1` instead of `6`?

juliusberner replied:
It is also used to initialize the `Sine` activation scaling, is this intentional?

juliusberner commented on neuralop/layers/embeddings.py:
```diff
@@ -73,4 +74,170 @@ def apply_2d_rotary_pos_emb(t, freqs_x, freqs_y):
+
+        weight = torch.zeros(dim_out, dim_in)
+        bias = torch.zeros(dim_out) if use_bias else None
+        self.init_(weight, bias, c=c, w0=w0)
```
Is there a reason that we do not use `torch.nn.Linear` and `torch.nn.init.uniform_` to initialize?

juliusberner commented on neuralop/models/base_model.py:
These edits could be removed from the PR

juliusberner commented on neuralop/layers/embeddings.py:
```diff
@@ -73,4 +74,170 @@ def apply_2d_rotary_pos_emb(t, freqs_x, freqs_y):
+        return out
+
+
+class SirenNet(nn.Module):
```
@dhpitt Should we move this to `layers/mlp.py`?

dhpitt replied:
That makes sense to me

juliusberner commented on neuralop/layers/embeddings.py:
```diff
@@ -73,4 +74,170 @@ def apply_2d_rotary_pos_emb(t, freqs_x, freqs_y):
+                          apply_rotary_pos_emb(t_y, freqs_y)), dim=-1)
+
+
+class GaussianFourierFeatureTransform(nn.Module):
```
@dhpitt We now have two Fourier feature embeddings, i.e., `PositionalEmbedding` above and this one --- I think we should name them more consistently, e.g., `GaussianFourierEmbedding` and `FourierEmbedding`?

dhpitt replied:
Good idea @zijieli-Jlee

dhpitt reviewed (approved):
This all looks great. @JeanKossaifi

juliusberner commented on neuralop/layers/embeddings.py:
```diff
@@ -73,4 +74,50 @@ def apply_2d_rotary_pos_emb(t, freqs_x, freqs_y):
+        # From [B, N, C] to [(B*N), C].
+        x = x.view(-1, channels)
+
+        x = x @ self._B.to(x.device)
```
We can remove the `.to(...)` since `self._B` is registered as parameter and automatically transferred with the model.

juliusberner commented on scripts/train_darcy_transformer.py:
```diff
@@ -0,0 +1,210 @@
+
+model = get_model(config)
+
+### The following wrapper is to maintain a consistent input format with other FNO-based model
```
Is there a reason why we do not directly use the FNO input format for the `TransformerNO`?

JeanKossaifi replied:
@dhpitt seems like we'd want to use the refactored 2DGrid here instead.

juliusberner reviewed (approved):
LGTM, I just added two minor comments.

JeanKossaifi commented:
@zijieli-Jlee this looks great, I agree with @juliusberner it would just be great to make the input the same as FNO and obviate the need for the wrapper :) 
We can also use the positional embedding already existing in the lib!

dhpitt commented:
Just resolved merge conflicts with changes in `neuralop.layers.embeddings`

juliusberner commented on neuralop/layers/transformer_block.py:
```diff
@@ -0,0 +1,322 @@
+                u: torch.Tensor, input tensor of shape [batch_size, num_src_grid_points, channels]
+                pos_src: torch.Tensor, grid point coordinates of shape [batch_size, num_src_grid_points, channels]
+                pos_emb_module: nn.Module, positional embedding module, by default None
+                pos_qry: torch.Tensor, grid point coordinates of shape [batch_size, num_sry_grid_points, channels],
```
nit: `sry` -> `qry`

dhpitt commented:
Converting this PR to draft until next release.

dhpitt commented:
Stale; closing
